### PR TITLE
[feature-wip](array-type) array_contains support more nested data types

### PR DESCRIPTION
**Pull request by cambyzju**
# Proposed changes

Issue Number: close #7570

## Problem Summary:
1. array_contains, array_position and element_at support more nested data types;
2. refractor FunctionArrayIndex;
3. array functions support all const arguments, such as:
> select array_position([33333,1,2],2);
```
+---------------------------------------+
| array_position(ARRAY(33333, 1, 2), 2) |
+---------------------------------------+
|                                     3 |
+---------------------------------------+
```

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
5. Has unit tests been added: (Yes/No/No Need)
6. Has document been added or modified: (Yes/No/No Need)
7. Does it need to update dependencies: (Yes/No)
8. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: gensrc/script/doris_builtins_functions.py [hunk 1]
```diff
@@ -117,67 +117,45 @@
...
+    [['array_position'], 'BIGINT', ['ARRAY_TINYINT', 'TINYINT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_SMALLINT', 'SMALLINT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_INT', 'INT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_BIGINT', 'BIGINT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_LARGEINT', 'LARGEINT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_DATETIME', 'DATETIME'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_DATE', 'DATE'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_FLOAT', 'FLOAT'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_DOUBLE', 'DOUBLE'], '', '', '', 'vec', ''],
+    [['array_position'], 'BIGINT', ['ARRAY_DECIMALV2', 'DECIMALV2'], '', '', '', 'vec', ''],
```
Comment (HappenLee) [inline gensrc/script/doris_builtins_functions.py:156]: the return result type of `array_position` of 'element_at' param type should be same

Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:157]: done

element_at param type has changed to BIGINT, keep the same with the result type of array_position.

File: be/src/vec/functions/array/function_array_index.h [hunk 2]
```diff
@@ -213,24 +139,131 @@ class FunctionArrayIndex : public IFunction {
...
     Status _execute_non_nullable(Block& block, const ColumnNumbers& arguments, size_t result,
                                  size_t input_rows_count) {
-        WhichDataType right_type(block.get_by_position(arguments[1]).type);
-        if ((right_type.is_string() &&
-             _execute_string(block, arguments, result, input_rows_count)) ||
-            _execute_number<NUMBER_TPL_PACK>(block, arguments, result, input_rows_count)) {
+        // extract array offsets and nested data
+        auto left_column =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
+        const auto array_column = check_and_get_column<ColumnArray>(*left_column);
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_index.h:191]: same to up

File: be/src/vec/functions/array/function_array_element.h [hunk 3]
```diff
@@ -181,7 +183,8 @@ class FunctionArrayElement : public IFunction {
                                     const DataTypePtr& result_type, size_t input_rows_count,
                                     const UInt8* src_null_map, UInt8* dst_null_map) {
         // check array nested column type and get data
-        auto array_column = check_and_get_column<ColumnArray>(*arguments[0].column);
+        auto left_column = arguments[0].column->convert_to_full_column_if_const();
+        const auto array_column = check_and_get_column<ColumnArray>(*left_column);
         DCHECK(array_column != nullptr);
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/functions/array/function_array_element.h]: done

File: be/src/vec/functions/array/function_array_index.h (hunk 2 above)
Comment (cambyzju) [reply be/src/vec/functions/array/function_array_index.h]: done

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**